File: fe/admin/src/services/authService.js
/**
 * Authentication Service
 */
import api from './api';

const authService = {
  /**
   * Login user
   * @param {string} username
   * @param {string} password
   * @returns {Promise<{access_token, user}>}
   */
  async login(username, password) {
    const response = await api.post('/auth/login', {
      username,
      password,
    });

    const { access_token, user } = response.data;

    // Store token and user in localStorage
    localStorage.setItem('access_token', access_token);
    localStorage.setItem('user', JSON.stringify(user));

    return response.data;
  },

  /**
   * Register new user
   * @param {Object} userData
   * @returns {Promise<User>}
   */
  async register(userData) {
    const response = await api.post('/auth/register', userData);
    return response.data;
  },

  /**
   * Logout user
   */
  async logout() {
    try {
      await api.post('/auth/logout');
    } catch (error) {
      console.error('Logout error:', error);
    } finally {
      // Clear local storage
      localStorage.removeItem('access_token');
      localStorage.removeItem('user');
    }
  },

  /**
   * Get current user info
   * @returns {Promise<User>}
   */
  async getCurrentUser() {
    const response = await api.get('/auth/me');
    return response.data;
  },

  /**
   * Verify token
   * @returns {Promise<boolean>}
   */
  async verifyToken() {
    try {
      const response = await api.get('/auth/verify');
      return response.data.valid;
    } catch (error) {
      return false;
    }
  },

  /**
   * Get stored user from localStorage
   * @returns {User|null}
   */
  getStoredUser() {
    const userStr = localStorage.getItem('user');
    return userStr ? JSON.parse(userStr) : null;
  },

  /**
   * Get stored token from localStorage
   * @returns {string|null}
   */
  getStoredToken() {
    return localStorage.getItem('access_token');
  },

  /**
   * Check if user is authenticated
   * @returns {boolean}
   */
  isAuthenticated() {
    return !!this.getStoredToken();
  },
};

export default authService;

File: be/services/auth_service.py
"""
Authentication service
"""
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from datetime import datetime

from entities.user import User
from schemas.user import UserCreate, UserLogin, TokenResponse, UserResponse
from utils.security import verify_password, get_password_hash, create_access_token


class AuthService:
    """Authentication service for login/logout/register"""
    @staticmethod
    def verify_password(plain_password: str, hashed_password: str) -> bool:
        """Verify password wrapper"""
        return verify_password(plain_password, hashed_password)

    @staticmethod
    def get_password_hash(password: str) -> str:
        """Hash password wrapper"""
        return get_password_hash(password)
    @staticmethod
    def register_user(db: Session, user_data: UserCreate) -> User:
        """
        Register a new user
        Args:
            db: Database session
            user_data: User registration data
        Returns:
            Created user
        Raises:
            HTTPException: If username or email already exists
        """
        # Check if username exists
        existing_user = db.query(User).filter(User.username == user_data.username).first()
        if existing_user:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Username already registered"
            )

        # Check if email exists
        existing_email = db.query(User).filter(User.email == user_data.email).first()
        if existing_email:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Email already registered"
            )

        # Create new user
        hashed_password = get_password_hash(user_data.password)
        db_user = User(
            username=user_data.username,
            email=user_data.email,
            full_name=user_data.full_name,
            hashed_password=hashed_password,
            role=user_data.role,
            grade=user_data.grade,
            class_name=user_data.class_name
        )

        db.add(db_user)
        db.commit()
        db.refresh(db_user)

        return db_user

    @staticmethod
    def login(db: Session, credentials: UserLogin) -> TokenResponse:
        """
        Authenticate user and return token
        Args:
            db: Database session
            credentials: Login credentials
        Returns:
            Token response with user data
        Raises:
            HTTPException: If credentials are invalid
        """
        # Find user by username
        user = db.query(User).filter(User.username == credentials.username).first()

        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect username or password"
            )

        # Verify password
        if not verify_password(credentials.password, user.hashed_password):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect username or password"
            )

        # Check if user is active
        if not user.is_active:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Account is inactive"
            )

        # Update last login
        user.last_login = datetime.utcnow()
        db.commit()

        # Create access token
        access_token = create_access_token(
            data={
                "sub": str(user.id),
                "username": user.username,
                "role": user.role.value
            }
        )

        # Return token and user data
        return TokenResponse(
            access_token=access_token,
            token_type="bearer",
            user=UserResponse.from_orm(user)
        )

    @staticmethod
    def get_current_user(db: Session, user_id: int) -> User:
        """
        Get current authenticated user
        Args:
            db: Database session
            user_id: User ID from token
        Returns:
            User object
        Raises:
            HTTPException: If user not found
        """
        user = db.query(User).filter(User.id == user_id).first()

        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )

        if not user.is_active:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Account is inactive"
            )

        return user
